### PR TITLE
x/gamm: Add PoolI mocks and test for inactive pools

### PR DESCRIPTION
**Pull request by rrrliu**
Closes: #2062

## What is the purpose of the change

Following up to #2050, add tests to ensure swaps fail when `IsActive` returns false. Requires mocking out the `PoolI` interface, so used `gomock` package for as per suggestion in https://github.com/osmosis-labs/osmosis/issues/1904#issuecomment-1182730437.

## Brief Changelog

- `gomock` package in `go.mod`
- `mock_pool.go` module for mocked `PoolI` interface
- Added new test for inactive pool swaps in `swap_tests.go` 

## Testing and Verifying

This change added tests and can be verified as follows:

- Extended `swap_tests.go` by adding `TestInactivePoolFreezeSwaps`

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (no)
  - How is the feature or change documented? (not applicable, also generated mock files are somewhat self-documenting)

### CODE REVIEW

File: tests/mocks/mock_pool.go
Comment (mattverse) [inline tests/mocks/mock_pool.go:1]: Oh interesting, is there a specific command to generate mock file for a specific file?

File: x/gamm/keeper/swap_test.go
Comment (mattverse) [inline x/gamm/keeper/swap_test.go:288]:
```suggestion
	gammKeeper := suite.App.GAMMKeeper
```

File: x/gamm/keeper/swap_test.go
Comment (mattverse) [inline x/gamm/keeper/swap_test.go:292]:
```suggestion
	inactivePoolId := gammKeeper.GetNextPoolNumberAndIncrement(suite.Ctx)
```

File: x/gamm/keeper/swap_test.go
Comment (mattverse) [inline x/gamm/keeper/swap_test.go:295]:
```suggestion
	gammKeeper.SetPool(suite.Ctx, inactivePool)
```

File: x/gamm/keeper/swap_test.go
Comment (mattverse) [inline x/gamm/keeper/swap_test.go:308]:
```suggestion
		_, swapInErr := gammKeeper.SwapExactAmountIn(suite.Ctx, suite.TestAccs[0], tc.poolId, testCoin, "bar", sdk.ZeroInt())
```

File: x/gamm/keeper/swap_test.go
Comment (mattverse) [inline x/gamm/keeper/swap_test.go:309]:
```suggestion
		_, swapOutErr := gammKeeper.SwapExactAmountOut(suite.Ctx, suite.TestAccs[0], tc.poolId, "bar", sdk.NewInt(1000000000000000000), testCoin)
```

File: x/gamm/keeper/swap_test.go
Comment (mattverse) [inline x/gamm/keeper/swap_test.go:294]: what's the reason for this? Wouldnt it already return the correct pool Id by defauly anyways?

Comment (mattverse) [inline x/gamm/keeper/swap_test.go:294]: Consider adding comments for these! imo could be really helpful since we're not very famialiar with how mock package works! (Also consider adding more detailed comments on how we mock a pool in the comment for the test itself!)

**Review by mattverse — APPROVED** [review]
Sorry for the late review here!

Nice work done here using the mock package! I'm really impressed! Left some minor comments, please take a look! Other than that LGTM!

File: x/gamm/keeper/swap_test.go
Comment (rrrliu) [reply x/gamm/keeper/swap_test.go:294]: Ah this is because the `NewMockPoolI` object doesn't actually have implemented functions (like a `BalancerPool` object) -- rather all its function calls throw errors by default.

Because the swap functions use `GetId`, we have to explicitly use `Return` to ensure that it returns the correct value rather than throwing an error. Will add a comment.

File: tests/mocks/mock_pool.go
Comment (rrrliu) [reply tests/mocks/mock_pool.go:1]: Yes! The [mockgen CLI](https://github.com/golang/mock#running-mockgen) handles automatic generation of mock interfaces.

**Comment by rrrliu** [comment]
@mattverse thanks for the suggestions, addressed your comments! Lmk if you want me to add anything about `gomock` in any documentation.

**Review by czarcas7ic — APPROVED** [review]
This LGTM, there is slight deviance in table driven test format but shouldn't block this merge IMO. Nice work!

File: x/gamm/keeper/swap_test.go
Comment (p0mvn) [inline x/gamm/keeper/swap_test.go:285]:
```suggestion
	testCoin := sdk.NewCoin(sdk.DefaultBondDenom, sdk.NewInt(10))
```

File: x/gamm/keeper/swap_test.go
Comment (p0mvn) [inline x/gamm/keeper/swap_test.go:314]:
```
t.Run(...)
```

**Review by p0mvn — APPROVED** [review]
Nice work. The core logic of the test looks great. Would like to suggest some minor improvements

**Comment by rrrliu** [comment]
> Hi @rrrliu . Great work on this PR! Just wanted to check in and see if you have time to address the last 2 suggestions. We can merge this right after 🚀

Just pushed -- sorry for the long turnaround!

**Comment by p0mvn** [comment]
@rrrliu thanks and np! Seems that there is a CI failure due to merge conflict, do you mind fixing that test, please? We can merge this right after

File: x/gamm/keeper/swap_test.go
Comment (rrrliu) [reply x/gamm/keeper/swap_test.go:285]: test failed when with this error (denom not found) when i used `sdk.DefaultBondDenom` so I changed it back -- any leads on this?

```
--- FAIL: TestMsgServerTestSuite (4.21s)
    --- FAIL: TestMsgServerTestSuite/TestInactivePoolFreezeSwaps (0.06s)
        --- FAIL: TestMsgServerTestSuite/TestInactivePoolFreezeSwaps/swap_succeeds_on_active_pool (0.00s)
            swap_test.go:319: 
                	Error Trace:	/Users/richardliu/osmosis/x/gamm/keeper/swap_test.go:319
                	            				/Users/richardliu/osmosis/x/gamm/keeper/suite.go:91
                	Error:      	Received unexpected error:
                	            	one of the provided pool denoms does not exist in pool
                	Test:       	TestMsgServerTestSuite/TestInactivePoolFreezeSwaps/swap_succeeds_on_active_pool
```

**Pull request merged**
